src/concurrency/local_epoch.cpp

```diff
@@ -16,17 +16,29 @@
 namespace peloton {
 namespace concurrency {
 
-  bool LocalEpoch::EnterEpoch(const uint64_t epoch_id) {
+  bool LocalEpoch::EnterEpoch(const uint64_t epoch_id, const bool is_snapshot_read) {
+
     epoch_lock_.Lock();
+
     // if this thread is never used or GC'd
     if (epoch_lower_bound_ == UINT64_MAX) {
+
       epoch_lower_bound_ = epoch_id - 1;
+    
     } else if (epoch_lower_bound_ >= epoch_id) {
-    // epoch_lower_bound_ has already been updated by the GC.
-    // have to grab a newer epoch_id.
-      epoch_lock_.Unlock();
-      return false;
-    } 
+    
+      if (is_snapshot_read == true) {
+    
+        epoch_lower_bound_ = epoch_id - 1;
```

Comment:
@jessesleeping The local thread is unaware of the progress of any other threads. It is possible that thread `TH1` is already at epoch 100, but another thread `TH2` is still at epoch 10. In this case, the centralized epoch manager should tell `TH1` that the progress from a global view is still at epoch 10.